### PR TITLE
[WIP] Add typings

### PR DESCRIPTION
**Pull request by Skillz4Killz**
Hello 👋 

This PR adds in typings so that eris-sharder can be used by TS users. I decided to make a draft PR so that it would be easier to discuss regarding any of the types I was unaware of. I've tried to write `Map<>` or `unknown` for the places that I was unable to determine what the type should be.

Thank you

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  interface Sharder extends EventEmitter {
+    shardCount: number;
+    firstShardID: number;
+    lastShardID: number;
+    clusterCount: number;
+    clusterTimeout: number;
+    token: string | false;
+    clusters: Map<>;
+    workers: Map<>;
+    queue: ErisSharderQueue;
```
Comment (AnotherGenZ) [inline index.d.ts]: Not sure what type this is

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    eris: Client;
+    options: {
+      stats: ErisSharderStats;
+    };
+    statsInterval: number;
+    mainFile: unknown;
```
Comment (AnotherGenZ) [inline index.d.ts]: string

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    name: string;
+    guildsPerShard: number;
+    webhooks: {
+      cluster?: unknown;
```
Comment (AnotherGenZ) [inline index.d.ts]: string

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      shard?: unknown;
```
Comment (AnotherGenZ) [inline index.d.ts]: string

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    unregister(name: string): void;
+    broadcast(name: string, message?: unknown): void;
+    sendTo(cluster: number, name: string, message?: unknown): void;
+    fetchUser(id: string): Promise<User>;
+    fetchGuild(id: string): Promise<Guild>;
+    fetchChannel(id: string): Promise<Channel>;
+  }
+
+  interface ErisSharderEvent {
+    fn: (msg: unknown) => void;
```
Comment (AnotherGenZ) [inline index.d.ts]: Object

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    clusterID: number;
+    ipc: ErisSharderIPC;
+    restartCluster(clusterID: number): void;
+  }
+
+  interface ErisSharderIPC extends EventEmitter {
+    events: Map<string, ErisSharderEvent>;
+    register(event: string, callback: Function): void;
+    unregister(name: string): void;
+    broadcast(name: string, message?: unknown): void;
```
Comment (AnotherGenZ) [inline index.d.ts]: Object

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    sendTo(cluster: number, name: string, message?: unknown): void;
```
Comment (AnotherGenZ) [inline index.d.ts]: Object

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  interface ErisSharderStats {
+    stats: {
+      guilds: number;
+      users: number;
+      totalRam: number;
+      voice: number;
+      exclusiveGuilds: number;
+      largeGuilds: number;
+      clusters: unknown[];
```
Comment (AnotherGenZ) [inline index.d.ts]: Object[]

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+		launch(): void;
+		chunk(shards: unknown, clusterCount: number): unknown[]
+		connectShards(): void;
+		sendWebhook(type: string, embed): void;
+		printLogo(): void;
+		restartCluster(worker: unknown, code: unknown): void;
+		calculateShards(): Promise<unknown>;
+		fetchInfo(start: number, type: unknown, value: unknown): void;
+		broadcast(start: number, message: unknown): void;
+		sendTo(cluster: unknown, message: unknown): void;
```
Comment (AnotherGenZ) [inline index.d.ts]: Object

File: index.d.ts [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+		start(clusterID: number): void;
+		launch(): void;
+		chunk(shards: unknown, clusterCount: number): unknown[]
+		connectShards(): void;
+		sendWebhook(type: string, embed): void;
+		printLogo(): void;
+		restartCluster(worker: unknown, code: unknown): void;
+		calculateShards(): Promise<unknown>;
+		fetchInfo(start: number, type: unknown, value: unknown): void;
+		broadcast(start: number, message: unknown): void;
```
Comment (AnotherGenZ) [inline index.d.ts]: Object

File: index.d.ts [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+		startStats(): void;
+		executeStats(clusters: unknown[], start: number): void;
+		start(clusterID: number): void;
+		launch(): void;
+		chunk(shards: unknown, clusterCount: number): unknown[]
+		connectShards(): void;
+		sendWebhook(type: string, embed): void;
+		printLogo(): void;
+		restartCluster(worker: unknown, code: unknown): void;
+		calculateShards(): Promise<unknown>;
```
Comment (AnotherGenZ) [inline index.d.ts]: number

File: index.d.ts [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    webhooks: {
+      cluster?: unknown;
+      shard?: unknown;
+    };
+    debug: boolean;
+    clientOptions: ClientOptions;
+		callbacks: Map<>;
+		isMaster(): boolean;
+		startStats(): void;
+		executeStats(clusters: unknown[], start: number): void;
```
Comment (AnotherGenZ) [inline index.d.ts]: Object[]

File: index.d.ts [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+		start(clusterID: number): void;
+		launch(): void;
+		chunk(shards: unknown, clusterCount: number): unknown[]
```
Comment (AnotherGenZ) [inline index.d.ts]: number[], returns number[][]

File: index.d.ts [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+		connectShards(): void;
+		sendWebhook(type: string, embed): void;
+		printLogo(): void;
+		restartCluster(worker: unknown, code: unknown): void;
```
Comment (AnotherGenZ) [inline index.d.ts]: nodejs worker (should be a type for this somewhere), number

**Review by AnotherGenZ — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 1 above)
Comment (Skillz4Killz) [reply index.d.ts]: This is a placeholder for typings that I need to make for the Queue class in the utils folder.

File: index.d.ts (hunk 5 above)
Comment (Skillz4Killz) [reply index.d.ts]: Using `Object` type in TS is basically useless and will cause more harm than good. If you mean that it has to be some object like `{}` then the way to type it is to type the allowed values in that object in detail **OR** leave it as `unknown`.

Comment (Skillz4Killz) [reply index.d.ts]: This is the Object type:
```ts
interface Object {
  // ...

  /** Returns a string representation of an object. */
  toString(): string;

  /** Returns a date converted to a string using the current locale. */
  toLocaleString(): string;

  /** Returns the primitive value of the specified object. */
  valueOf(): Object;

  /**
   * Determines whether an object has a property with the specified name.
   * @param v A property name.
   */
  hasOwnProperty(v: string): boolean;

  /**
   * Determines whether an object exists in another object's prototype chain.
   * @param v Another object whose prototype chain is to be checked.
   */
  isPrototypeOf(v: Object): boolean;

  /**
   * Determines whether a specified property is enumerable.
   * @param v A property name.
   */
  propertyIsEnumerable(v: string): boolean;
}```

File: index.d.ts (hunk 14 above)
Comment (Skillz4Killz) [reply index.d.ts]: Couldn't find the nodejs worker type so i kept it as unknown for now.

**Comment by Rexogamer** [comment]
Just wondering why this was closed? We at DTel need these for v4, for example

**Comment by AnotherGenZ** [comment]
I'm not too sure why this was closed. I'd have to go back and check if the proposed typings are still up to date/accurate if I was to accept them.

**Pull request closed without merging**